### PR TITLE
Make PurchaseParams API & Win-Back Functions Public

### PR DESCRIPTION
**Pull request by fire-at-will**
### Description
This PR will ship the following functionality:
- PurchaseParams API
- Fetching & purchasing win-back offers on a developer's paywall

### Specific Changes
- Makes the `PurchaseParams` functionality publicly available by removing the `ENABLE_PURCHASE_PARAMS` compiler flag checks
- Makes the `PurchaseParams` functionality unavailable if the `ENABLE_CUSTOM_ENTITLEMENT_COMPUTATION` flag is used
- Make the functions to fetch & redeem eligible win-back offers public and exposed to Objective-C
- Makes the new transaction metadata functionality public
- Adds the new functions to the purchase testers
- Hides the transaction metadata feature behind a new `ENABLE_TRANSACTION_METADATA` compiler flag
- Changes the function signature of the completion handler-based function to fetch eligible win-back offers from
```swift
@available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
func eligibleWinBackOffers(
  forProduct product: StoreProduct,
  completion: @escaping @Sendable (Result<[WinBackOffer], PublicError>) -> Void
)
```
to
```swift
@available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
 func eligibleWinBackOffers(
     forProduct product: StoreProduct,
     completion: @escaping @Sendable ([WinBackOffer]?, PublicError?) -> Void
 )
```
So that the function can be called from Objective-C (and therefore the hybrid SDKs that rely on Objective-C)

### CODE REVIEW

**Comment by fire-at-will** [comment]
@RCGitBot please test

File: Sources/Misc/Concurrency/Purchases+async.swift [hunk 1]
```diff
@@ -83,6 +83,16 @@ extension Purchases {
...
 
+    func restorePurchasesAsync() async throws -> CustomerInfo {
+        return try await withUnsafeThrowingContinuation { continuation in
+            self.restorePurchases { customerInfo, error in
+                continuation.resume(with: Result(customerInfo, error))
+            }
+        }
+    }
+
+    #if !ENABLE_CUSTOM_ENTITLEMENT_COMPUTATION
```
Comment (fire-at-will) [inline Sources/Misc/Concurrency/Purchases+async.swift:94]: Git is rendering this diff oddly, essentially what I did here was move the `func purchaseAsync(_ params: PurchaseParams)` function down in the file beneath `restorePurchasesAsync()` so it'd be inside the `#if !ENABLE_CUSTOM_ENTITLEMENT_COMPUTATION` block

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Review by MarkVillacampa — APPROVED** [review]

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Review by MarkVillacampa — APPROVED** [review]

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Comment by fire-at-will** [comment]
@RCGitBot please test

**Pull request merged**